### PR TITLE
feat: v8

### PR DESCRIPTION
**Pull request by nickfloyd**
BREAKING CHANGE: Replace support for Node.js http(s) Agents with documentation on using fetch dispatchers instead

RE: https://github.com/octokit/octokit.js/discussions/2484

Node's native fetch method does not support using a Node http(s) agent, because it's not built on top of Node's http module (it's built on top of the net module instead).

----

### Before the change?

* Users of the @octokit/request module could pass in request options to fetch and request would map those

### After the change?

* Users should now implement a custom fetch function with a dispatcher preset

----

### Pull request checklist
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Added the appropriate label for the given change

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/main/community/breaking_changes.md) to help!

- [x] Yes (Please add the `Type: Breaking change` label)
- [ ] No

----

### CODE REVIEW

**Comment by wolfy1339** [comment]
This removes all options, which would be undesirable.

**Comment by nickfloyd** [comment]
@wolfy1339 Thanks for the introspection here.

@gr2m and I had a conversation about this change where three different scenarios presented themselves as to how to approach this (I'll paraphrase, and hopefully, @gr2m will jump in so that we can have a complete conversation about it).

Here are the 3 options:

1. Remove  `requestOptions.request as any` ([this](https://github.com/octokit/request.js/blob/main/src/fetch-wrapper.ts#L53-L54)) and fix the breaking references
2. Remove any options that we implement in our fetch wrapper that are web standards only - removing all of the ones based on node-fetch - specifically around the standards defined [here](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) and removing any node-fetch [here](https://github.com/node-fetch/node-fetch#fetch-options) specific options.
3. Remove all options passed and mapped in our fetch wrapper.  While this is the highest degree of breaking change, this does exactly what we need to decouple things. i.e. removing the Object.assign in the fetch-wrapper.ts - [here](https://github.com/octokit/request.js/blob/main/src/fetch-wrapper.ts#L41-L54) - like I've done in this PR.

What do you think about the above?  If you get a chance, please share why you think option three would be less desirable. 

I could definitely be missing something obvious as well, so your thoughts and perspective here would be appreciated!

**Comment by wolfy1339** [comment]
I believe that transparently passing options to the fetch function, would be ideal.

That way we don't do any specific handling on our side, and users don't have to use a custom fetch function when they need to pass options.

We can then remove documentation on the options, and direct users to the documentation for their implementation of that fetch function l, whether it be `node-fetch` (or another similar library) or the Fetch API.

That would offer some lesser level of breaking change for users

**Comment by nickfloyd** [comment]
Hey @wolfy1339, thanks for the thoughts/feedback - we're you thinking the implementation would look something like [this](https://github.com/octokit/request.js/pull/599/files#diff-e25dfcb05e38ef790c0894be934f090e2f095a8bf496aae32e908d7356b10c8cR39)?

**Comment by wolfy1339** [comment]
Changes need to be made to `@octokit/types` first before merging this, to remove the `node-fetch` specific options:
- `agent`
- `timeout` (can be replaced with docs on how to use the `signal` option with `AbortSignal`, see [the v3 upgrade guide of `node-fetch`](https://github.com/node-fetch/node-fetch/blob/main/docs/v3-UPGRADE-GUIDE.md#the-timeout-option-was-removed))

***

Currently, this won't work as-is.
The `requestOptions` contains all options for the request which is non-standard, and custom to our API.


Here is a couple points I'm seeing that will cause conflict with this PR:
- This would break all requests that need to use different HTTP methods as we don't pass this information along.
- For `POST`, or `PUT` requests, one needs to be able to pass a body to the HTTP request.
- Since we aren't passing along headers, which are needed for the `requestOptions.mediaType` options, those requests would default to use `json` instead of the intended mediatype format

File: test/request.test.ts
Comment (wolfy1339) [inline test/request.test.ts:565]: Linking docs for the `signal` option for clearer context: https://developer.mozilla.org/en-US/docs/Web/API/fetch#signal

This test seems like a sanity check to make sure the option is passed along by passing a value that isn't the expected type and checking if it throws an error

Example:

```
> fetch("https://example.com",{signal:"bar"})
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 294,
  [Symbol(trigger_async_id_symbol)]: 5
}
> Uncaught:
TypeError: Failed to construct 'Request': member signal is not of type AbortSignal.
    at Object.fetch (node:internal/deps/undici/undici:11457:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by wolfy1339** [comment]
Even with the most recent changes, this won't work.

Continuing on with the thoughts on my previous comment,

For option 3, (or is it option 1) only allow setting `headers`, `method`, `url`, `body`, `redirect` options, and the duplex option (as it is needed when there is a body) since those options are needed for downstream modules.

To pass through options, we would still require the `Object.assign()` there isn't a way around that, it would satisfy option 2

**Comment by gr2m** [comment]
how about we add `options.request.fetchOptions`, and `fetchOptions` can contain all fetch options that the native `fetch` supports? That way we don't need merge all of `options.request` into the fetch request call. And it would make it all explicit.

I was initially preferring to just support `options.request.fetch`, but I can now imagine that there will be valid use cases where we can't set `fetch`, but still want to pass fetch options such as a signal in order to abort a request. For example in the context of plugins, we cannot override the `fetch` used by the `octokit` instance, but we still might want to pass a `signal` on a per-request basis.

File: README.md
Comment (gr2m) [inline README.md:278]: wait why was method removed?

**Comment by wolfy1339** [comment]
> how about we add `options.request.fetchOptions`, and `fetchOptions` can contain all fetch options that the native `fetch` supports? That way we don't need merge all of `options.request` into the fetch request call. And it would make it all explicit.

:+1: This is a good idea

File: README.md
Comment (nickfloyd) [reply README.md:278]: It was not intentional/part of the final changeset, just inflight drafting...

**Comment by nickfloyd** [comment]
> how about we add options.request.fetchOptions, and fetchOptions can contain all fetch options that the native fetch supports? That way we don't need merge all of options.request into the fetch request call. And it would make it all explicit.

This seems like a reasonable approach. Concretely you are suggesting the following (please correct me if I am wrong on anything so that I know we are all on the same page):

1. Create a new option named `options.request.fetchOptions` which is the container for the explicit options from native `fetch` (according to [this](https://developer.mozilla.org/en-US/docs/Web/API/fetch)) which are:
- options.method ✅ 
- options.headers ✅ 
- options.body ✅ 
- options.mode ✅ 
- options.credentials❓ (**NOTE: supporting this in fetchOptions does not seem right given how we handle auth**)
 -  options.credentials.omit❓(should we support this?)
 -  options.credentials.same-origin❓(should we support this?)
 -  options.credentials.include❓(should we support this?)
- options.cache ✅ 
- options.redirect ✅ 
- options.referrer❓(should we support this?)
- options.referrerPolicy
- options.integrity❓(should we support this?)
- options.keepalive ✅ 
- options.signal ✅ 
- options.priority❓(should we support this?)

Some of the options above seem superfluous/unnecessary for the type of support we're trying to give, the way API requests are made to our API -  things like: `options.credentials.*`, `options.referrer`, `options.integrity`, and `options.priority`

2. Keep 
- `route`
- `options.baseUrl`
- `options.url` (mapped as Fetch `resource`)
- `options.mediaType.format`
- `options.request.log`
- `options.mediaType.previews` (since this library also allows interactions with GraphQL we have to keep this one in)
- `options.request.hook`

4. Remove the following options in the docs as well as how they are referenced/used in the source in favor of the above implementation:
- options.headers
- options.method
- options.data
- options.request.agent
- options.request.signal

5. Update any related or currently missing types that support these parameters in @octokit/types [here](https://github.com/octokit/types.ts/pull/561) and reference in the beta here.

Let me know if all of this sounds right or what needs to be adjusted.

**Comment by kfcampbell** [comment]
I think @nickfloyd's take is reasonable; we shouldn't support other ways of passing in authentication. I'm ready/willing to carry out this work if we can get consensus on the approach.

File: test/request.test.ts
Comment (kfcampbell) [reply test/request.test.ts:565]: In that case, with these breaking changes, should this test be removed? If we're expecting the caller to bring their own fetch, perhaps we shouldn't test on the internals of those options.

**Comment by kfcampbell** [comment]
Here's a follow-up based on Nick's plan above: 

1. Create a new option named `options.request.fetchOptions` which is the container for the following subset of the explicit options from native `fetch` (according to [this](https://developer.mozilla.org/en-US/docs/Web/API/fetch)):
- options.method ✅ 
- options.headers ✅ 
- options.body ✅ 
- options.signal ✅ 

3. Keep 
- `route`
- `options.baseUrl`
- `options.url` (mapped as Fetch `resource`)
- `options.mediaType.format`
- `options.request.log`
- `options.mediaType.previews` (since this library also allows interactions with GraphQL we have to keep this one in)
- `options.request.hook`

4. Remove the following options in the docs as well as how they are referenced/used in the source in favor of the above implementation:
- options.headers
- options.method
- options.data
- options.request.agent
- options.request.signal

5. Update any related or currently missing types that support these parameters in @octokit/types [here](https://github.com/octokit/types.ts/pull/561) and reference in the beta here.

Thoughts?

**Comment by wolfy1339** [comment]
I agree with the plan. It seems well thought out

**Comment by gr2m** [comment]
> 5\. Remove the following options in the docs as well as how they are referenced/used in the source in favor of the above implementation:
> 
> * options.headers
> * options.method
> * options.data

Please don't remove these three, but implement them as alias and log a deprecation warning (once). But we can remove documentation

**Comment by kfcampbell** [comment]
@gr2m I've updated the plan accordingly.

File: README.md
Comment (wolfy1339) [inline README.md:276]: The docs should point that this option is for the GraphQL API only, and doesn't have an effect on REST routes

File: test/request.test.ts
Comment (wolfy1339) [reply test/request.test.ts:565]: It seems that this is the only test that checks if the options were properly passed along.

I think it should be fine to remove it

**Comment by kfcampbell** [comment]
I don't fully understand yet, so I have a few dumb questions before starting to drive this:

I'm looking at creating a type in fetch-wrapper.ts like:

```export type FetchOptions = {
  method: string;
  headers: { [header: string]: string };
  body?: string;
  signal?: AbortSignal;
};```

When we say that we want to keep "route", "options.baseURL", etc., should those be added to the above container type? These are things that seem like they belong in octokit/types.ts's EndpointInterface or RequestOptions files, and I'm not quite sure where the delineation should be.

**Comment by gr2m** [comment]
Hmm right I forgot about `@octokit/endpoint`. The idea for that library is to have a generic library that you can use with your own request method. I don't think `@octokit/endpoint` should use `fetchOptions` for `method`, `headers`, and `data`.

Maybe we should leave these untouched for now, and just start by removing `options.request.{agent,signal}` `options.fetchOptions.signal`? 🤔 Now I'm thinking we should even keep `options.request.signal` and just remove `options.request.agent` for now. Sorry for the back-and-forth, my head is in other places right now, and I'm not 100% sure what I think the best path forward is

**Comment by kfcampbell** [comment]
> Sorry for the back-and-forth

No worries at all, this is a complicated issue and I'm definitely not grasping it yet. 

I'd like to think about this a little more and reread all the content on this page. I'll probably come knocking at your door with some questions tomorrow if that's okay.

File: README.md
Comment (kfcampbell) [reply README.md:276]: I've updated the doc here to reflect that! Thanks.

File: README.md
Comment (wolfy1339) [inline README.md:276]:
```suggestion
      Name of previews, such as `mercy`, `symmetra`, or `scarlet-witch`. See <a href="https://docs.github.com/graphql/overview/schema-previews">GraphQL Schema Previews</a>.
```

**Comment by kfcampbell** [comment]
I've chatted with @gr2m today and he's recommended explicitly passing in signal without passing in the rest of the request options, so I've done so in https://github.com/octokit/request.js/pull/599/commits/35a84c6e84044801fad2176eb7be801af4e71ac2, which looks like:

`return fetch(requestOptions.url, { signal: (requestOptions as any).signal})`

Above, we had discussed doing the same for method, headers, and body. Additionally, we had discussed keeping data as well. I'd like to invite additional comments about that decision. In the morning, absent other inquiries, I'll adjust this PR accordingly.

**Comment by wolfy1339** [comment]
> Above, we had discussed doing the same for method, headers, and body. Additionally, we had discussed keeping data as well. I'd like to invite additional comments about that decision. In the morning, absent other inquiries, I'll adjust this PR accordingly.

There isn't much to be said, those options are needed for functionality downstream.

- headers options are needed for mediatype and GraphQL previews
- method option required for making POST/PUT/DELETE requests
- body is needed to pass along the data for the request to the API

Additionally, we should keep the duplex option as it is defined currently in `main`

File: src/fetch-wrapper.ts
Comment (wolfy1339) [inline src/fetch-wrapper.ts:41]: Is the `Object.assign()` necessary here? Could we not simply use an object

Comment (kfcampbell) [reply src/fetch-wrapper.ts]: You're correct, I've done so now.

**Comment by kfcampbell** [comment]
Step one in the plan presented [here](https://github.com/octokit/request.js/pull/599#issuecomment-1622454701) exists on this PR in its current state. Should the other steps be part of this PR as well or split up (perhaps using other PRs into this branch)?

**Comment by wolfy1339** [comment]
I think that for now, we can leave this PR as-is in order to unblock downstream updates ASAP.

Further follow-ups can be made for the remaining points

**Comment by kfcampbell** [comment]
Sounds good! I've marked the PR as ready for review. Since @nickfloyd and I were the committers, I'll wait for approval/merge from you or @gr2m before proceeding with rolling out updates across other repos.

**Comment by wolfy1339** [comment]
https://github.com/octokit/types.ts/pull/561 needs to be merged as a prerequisite to this PR

The version will need to be updated in this PR as well

**Comment by kfcampbell** [comment]
:+1: I've marked that PR as ready for review and I'll update the version here once it's released.

**Comment by wolfy1339** [comment]
That PR has been merged

**Comment by kfcampbell** [comment]
Thanks! Package has been updated here. After this, according to [the plan](https://github.com/octokit/octokit.js/issues/2486#issue-1769756358), I believe the idea is to update core.js first, then octokit.js, app.js, and rest.js. Is that correct?

**Comment by wolfy1339** [comment]
We need to update the dependencies of `@octokit/core` first, and so on.

Then the dependencies of `@octokit/rest`
Then dependencies of `@octokit/app`

Octokit.JS is the final one to get updated, as it depends on all the other modules

**Review by wolfy1339 — APPROVED** [review]

File: src/fetch-wrapper.ts
Comment (gr2m) [inline src/fetch-wrapper.ts:44]: Did you mean to pass `data` here? Where does that come from?

Comment (kfcampbell) [reply src/fetch-wrapper.ts:44]: @gr2m I was going off of https://github.com/octokit/request.js/pull/599#issuecomment-1622546734, though looking at it now I forgot the piece about aliasing.

Comment (kfcampbell) [reply src/fetch-wrapper.ts:44]: @gr2m should @wolfy1339 and I hold off on merging upstream bumps until a new major version has been cut? What do you think would be the best way forward here?

Comment (gr2m) [reply src/fetch-wrapper.ts:44]: It's okay it's just a fix. `options.data` is handled by `@octokit/endpoint`. There is no `options.request.data`. Sorry for the confusion.

Comment (kfcampbell) [reply src/fetch-wrapper.ts:44]: No problem! It's brought on by my lack of understanding. I've opened https://github.com/octokit/request.js/pull/604 to hopefully correct this.

**Pull request merged**
